simpleline/render/widget_containers.py

```diff
@@ -0,0 +1,349 @@
+        """Render all items and then find how many lines are required for every row.
+
+        This will call `self._render_all_items()` and `self._lines_per_every_row()` in correct order.
+        """
+        self._render_all_items()
+        return self._lines_per_every_row(ordered_items)
+
+    def _render_all_items(self):
+        for item_id, item in enumerate(self._widgets):
+            item_width = self._columns_width
+
+            if self._key_pattern:
+                # render numbers before widgets
+                number_widget = TextWidget(self._key_pattern.get_widget_key_from_id(item_id))
+                number_width = len(number_widget.text)
+                number_widget.render(number_width)
+                self._numbering_widgets.append(number_widget)
+                # reduce the size of widget because of the number
+                item_width -= number_width
+
+            item.widget.render(item_width)
+
+    def _lines_per_every_row(self, items):
+        # call `self._render_and_calculate_lines_per_rows()` method instead
+        lines_per_row = []
+
+        # go through all items and find how many lines we need for each row printed (because of wrapping)
+        for column_items in items:
+            for row_id, item_id in enumerate(column_items):
+                item = self._widgets[item_id]
+                if len(lines_per_row) <= row_id:
+                    lines_per_row.append(0)
+
+                lines_per_row[row_id] = max(lines_per_row[row_id], len(item.widget.get_lines()))
+
+        return lines_per_row
+
+    def _get_ordered_map(self):
+        """Return list of identifiers (index) to the original item list.
+
+        .. NOTE: Use of ``self._prepare_list()` is encouraged to create output list and just fill up this list.
+        """
+        # create list of columns (lists)
+        ordering_map = self._prepare_list()
+
+        for item_id in range(self.size):
+            ordering_map[item_id % self._columns].append(item_id)
+
+        return ordering_map
+
+    def _prepare_list(self):
+        """Prepare list for items ordering to rows and columns.
+
+        List will be prepared as ([column 1], [column 2], ...)
+        """
+        return list(map(lambda x: [], range(0, self._columns)))
+
+
+class ListColumnWidget(ListRowWidget):
+    """Place widgets in columns automatically.
+
+    Compared to the ColumnWidget this is able to handle word wrapping correctly.
+
+    Widgets will be placed based on the number of columns in the following way:
+
+    w1   w4   w7
+    w2   w5   w8
+    w3   w6   w9
+    """
+
+    def _get_ordered_map(self):
+        ordering_map = self._prepare_list()
+        items_in_column = ceil(len(self._widgets) / self._columns)
+
+        for item_id in range(self.size):
+            col_position = int(item_id // items_in_column)
+            ordering_map[col_position].append(item_id)
+
+        return ordering_map
+
+
+class KeyPattern(object):
```

Comment:
I don't agree with you. Patterns are nice feature but it is not enough in all cases. For example how would you specify numbering by Roman numerals or alphabet number?

That is why I wanted to have greater freedom on specifying how the numbering will be created. In worst case scenario, the user can create array with labels which will be used.

I agree that the methods have weird names, I wasn't able to find something better :( .